calico_containers/pycalico/ipam.py

```diff
@@ -15,78 +15,472 @@
 from etcd import EtcdKeyNotFound, EtcdAlreadyExist
 
 from netaddr import IPAddress, IPNetwork
+from types import NoneType
+import socket
+import json
+import logging
 
 from pycalico.datastore_datatypes import IPPool
 from pycalico.datastore import CALICO_V_PATH, DatastoreClient, handle_errors
+from pycalico.block import (AllocationBlock,
+                            get_block_cidr_for_address,
+                            BLOCK_PREFIXLEN,
+                            AlreadyAssignedError)
 
-IP_VERSION_PATH = CALICO_V_PATH + "/ipam/v%(version)s"
-IP_ASSIGNMENT_PATH = IP_VERSION_PATH + "/assignment/%(pool)s"
-IP_ASSIGNMENT_KEY = IP_ASSIGNMENT_PATH + "/%(address)s"
+_log = logging.getLogger(__name__)
 
+IPAM_V_PATH = "/calico/ipam/v1/"
+IPAM_HOST_PATH = IPAM_V_PATH + "host/%(hostname)s/"
+IPAM_HOST_AFFINITY_PATH = IPAM_HOST_PATH + "ipv%(version)d/block/"
+IPAM_BLOCK_PATH = IPAM_V_PATH + "assignment/ipv%(version)d/block/"
+RETRIES = 100
 
-class SequentialAssignment(object):
+my_hostname = socket.gethostname()
+
+
+class BlockReaderWriter(DatastoreClient):
     """
-    Assign IP addresses sequentially
+    Can read and write allocation blocks to the data store, as well as related
+    bits of state.
+
+    This class keeps etcd specific code from being in the main IPAMClient
+    class.
     """
 
-    def __init__(self):
-        # Init an etcd client.
-        self.etcd = IPAMClient()
+    def _read_block(self, block_cidr):
+        """
+        Read the block from the data store.
+        :param block_cidr: The IPNetwork identifier for a block.
+        :return: An AllocationBlock object
+        """
+        key = _datastore_key(block_cidr)
+        try:
+            result = self.etcd_client.read(key)
+        except EtcdKeyNotFound:
+            raise KeyError(str(block_cidr))
+        block = AllocationBlock.from_etcd_result(result)
+        return block
 
-    def allocate(self, pool):
+    def _compare_and_swap_block(self, block):
         """
-        Attempt to allocate an IP address from the provided pool.
+        Write the block using an atomic Compare-and-swap.
+        """
+
+        # If the block has a db_result, CAS against that.
+        if block.db_result is not None:
+            try:
+                self.etcd_client.update(block.update_result())
+            except EtcdAlreadyExist:
+                raise CASError(str(block.cidr))
+        else:
+            # Block is new.  Write it with prevExist=False
+            key = _datastore_key(block.cidr)
+            value = block.to_json()
+            try:
+                self.etcd_client.write(key, value, prevExist=False)
```

Comment:
Do we need to update db_result here - or does the API no allow us to re-use an AllocationBlock object without first re-querying it?